x/programs/cmd/simulator/src/lib.rs

```diff
@@ -193,20 +200,25 @@ pub struct PlanResult {
     pub response: Option<Vec<i64>>,
 }
 
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ClientError {
+    #[error("read")]
+    Read(#[from] std::io::Error),
+    #[error("serde_json error")]
+    Deserialization(#[from] serde_json::Error),
+    #[error("custom")]
+    Custom(String),
```

Comment:
This isn't a great practice. You're allocating a string for the error but not using it in the display of the error (only when debugging). It's better not to have "custom" errors but to have well-defined errors (or at least display the custom error).

I would change this to
 ```suggestion
    #[error("EOF")]
    Eof,
```

Reply:
Indeed, that's much better now thanks !